### PR TITLE
[Android] Implement ScrollView sticky headers on Android

### PR DESCRIPTION
**Pull request by janicduplessis**
This adds support for sticky headers on Android. The implementation if based primarily on the iOS one (https://github.com/facebook/react-native/blob/master/React/Views/RCTScrollView.m#L272) and adds some stuff that was missing to be able to handle z-index, view clipping, view hierarchy optimization and touch handling properly.

Some notable changes:
- Add `ChildDrawingOrderDelegate` interface to allow changing the `ViewGroup` drawing order using `ViewGroup#getChildDrawingOrder`. This is used to change the content view drawing order to make sure headers are drawn over the other cells. Right now I'm only reversing the drawing order as drawing only the header views last added a lot of complexity especially because of view clipping and I don't think it should cause issues.

- Add `collapsableChildren` prop that works like `collapsable` but applies to every child of the view. This is needed to be able to reference sticky headers by their indices otherwise some subviews can get optimized out and break indexes.

- Make `TouchTargetHelper` take drawing order into consideration when finding the touch target.

- Take translate into consideration when clipping subviews.

**Test plan**
Tested using the UIExplorer main menu on android.
Tested that touches get dispatched to the sticky header when over other touchable views.
Tested that touchables inside a sticky header work properly.

### CODE REVIEW

**Comment by facebook-github-bot-1** [comment]
By analyzing the blame information on this pull request, we identified @astreet and @rigdern to be potential reviewers.

**Comment by satya164** [comment]
cc @brentvatne :D

**Comment by brentvatne** [comment]
@janicduplessis - tried it out, seems great! any idea about the CI failures?

**Comment by janicduplessis** [comment]
Looks like the instrumentation tests are failing, I'll run them locally to see what is going on.

**Comment by sahrens** [comment]
Did you consider implementing this with the new native-driven Animated.event? Seems like it would be a little awkward to describe with Animated, but doable.

**Comment by brentvatne** [comment]
> Did you consider implementing this with the new native-driven Animated.event? Seems like it would be a little awkward to describe with Animated, but doable.

@sahrens - probably doable, but I think we should wait for offload Animated and Animated event stuff to be more mature first, then switch the iOS and Android implementations of sticky headers at the same time.

**Comment by nihgwu** [comment]
bravo

**Comment by janicduplessis** [comment]
@sahrens I agree with @brentvatne, it would be really nice to eventually convert it but for now I think this is better.

File: ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollView.java [hunk 1]
```diff
@@ -152,6 +175,22 @@ protected void onScrollChanged(int x, int y, int oldX, int oldY) {
   }
 
   @Override
+  public View hitTest(float[] coordinates) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollView.java]: Is this needed because our touch handling doesn't respect drawing order? If so, can we just make our normal touch handling handle drawing order so that we don't have to add a new touch handling code path?

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java [hunk 2]
```diff
@@ -289,7 +294,11 @@ private void updateClippingToRect(Rect clippingRect) {
 
   private void updateSubviewClipStatus(Rect clippingRect, int idx, int clippedSoFar) {
     View child = Assertions.assertNotNull(mAllChildren)[idx];
-    sHelperRect.set(child.getLeft(), child.getTop(), child.getRight(), child.getBottom());
+    sHelperRect.set(
+      child.getLeft() + (int) child.getTranslationX(),
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: This might not be the right way to round: what does AOSP do here, do you know?

**Comment by astreet** [comment]
Re collapsableChildren vs cloning for the collapsable prop: is there actually a perf concern here? If possible, I'd really rather not add another prop we need to support.

File: ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollView.java (hunk 1 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollView.java]: Yea, good idea, I just checked and AOSP does use getChildDrawingOrder when dispatching a touch event to subviews.

**Comment by facebook-github-bot-0** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/7767c127def0584f4292a3387fa83eb345450d63..11e9aecc982a820ff98990edd82afe0b6ffaf848)

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/7f8275acfd1b5dc4c09352fbf27e6b40ee320932..09655e6a9ca80ea7e12ca270d40f6f584ecfa61a)

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/09655e6a9ca80ea7e12ca270d40f6f584ecfa61a..886b17e47acceec145a5b92ed48d5a5b5c2e1d3b)

**Comment by janicduplessis** [comment]
@astreet Yea, unfounded perf concerns :), changed it to cloning child elements in JS.

**Comment by facebook-github-bot-8** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/886b17e47acceec145a5b92ed48d5a5b5c2e1d3b..aed01aeeb0eff3a3e512410b7f18f41f2447cb54)

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java (hunk 2 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Updated to do like AOSP, use Matrix#mapRect and take a sligtly larger rect when converting from a float rect to a int rect. This make sure that float -> int doesn't remove views that may still be visible.

https://github.com/android/platform_frameworks_base/blob/4535e11fb7010f2b104d3f8b3954407b9f330e0f/core/java/android/view/ViewGroup.java#L5049

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Actually I changed it to (floor, floor, ceil, ceil) that way it keeps the same behaviour for non decimal numbers and will still make sure to never clip views that end up with decimal bounds after transform.

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/aed01aeeb0eff3a3e512410b7f18f41f2447cb54..c7eefb9e3bbcc7c850a46124141b23b1c9ef7f5e)

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java [hunk 3]
```diff
@@ -437,7 +437,7 @@ private void transitionLayoutOnlyViewToNativeView(
 
   private static boolean isLayoutOnlyAndCollapsable(@Nullable ReactStylesDiffMap props) {
     if (props == null) {
-      return true;
+      return false;
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Why this change?

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Yea I was going to comment on this, If I don't do this native view counts doesn't match sticky indices. I think it is because of `null` react elements. Any other idea other than doing this?

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java [hunk 4]
```diff
@@ -336,8 +350,15 @@ private void updateSubviewClipStatus(View subview) {
 
     // do fast check whether intersect state changed
     sHelperRect.set(subview.getLeft(), subview.getTop(), subview.getRight(), subview.getBottom());
-    boolean intersects = mClippingRect
-        .intersects(sHelperRect.left, sHelperRect.top, sHelperRect.right, sHelperRect.bottom);
+    Matrix matrix = subview.getMatrix();
+    if (!matrix.isIdentity()) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: So, I realized that if the translation/rotation/etc are updated after this happens, clipping will no longer be correct. It's probably better to not factor in translation/etc until we have a way to solve that (if we ever do: it doesn't seem trivial to me :( )

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java (hunk 3 above)
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: But we're setting collapsable=false on all those children though -- I don't understand why that wouldn't be enough.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Some children are null so we can't set collapsable = false on them. I had to add a null check in children map here https://github.com/facebook/react-native/pull/9456/commits/1e87a600e6ba71d7bdcac2c02659988d4f3ddafb#diff-f8f8b220fc0d1573e4f8b78c84415075R455

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java (hunk 4 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Maybe disable clipping if the view has a transform? Because clipping + transform is broken at the moment anyway.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: We could also call `updateClippingRect` after updating the transform prop. This will cause a bit of extra work but will make sure that whatever happens last (setting transform or updating layout) will calculate the right clipping.

Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Are we actually hitting this right now?

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Yea, with sticky headers we translate them so they stay on the top of the scrollview when normally they should be outside of clip bounds without the transform.

Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: mm, right. When you say 
>We could also call updateClippingRect after updating the transform prop.
How exactly would that work?

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: I was thinking of something like
``` java
if (view instanceof ReactClippingViewGroup) {
  ((ReactClippingViewGroup) view).updateClippingRect();
}
```
in https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java#L53

Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Yeah, that sounds reasonable if you can get it to work. The best other solution I can think of is just not clipping transformed views no matter what, but I'm worried some developer is going to translate images (e.g. SoundCloud's android app as you scroll the main screen) and have a bad time.

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/c7eefb9e3bbcc7c850a46124141b23b1c9ef7f5e..8c39c9b73f931c881311664fe6ea1d7543d306d1)

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java (hunk 3 above)
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: I see. Then we should either have JS send the tags of the children that are sticky headers which the native view could then convert into indices (which would also fix the whole collapsable problem), or less ideally remap the indices in JS.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: How would you go about getting the sticky headers tags in JS? Seems like it would be relatively complex since we can only get them after the view has been rendered using refs. Maybe remapping indexes to offset for null views that will be removed would end up a lot simpler.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Ok this is actually harder than I thought because just adding `collapsable=false` doesn't really work because children are not actually `View` components so adding the `collapsable` prop doesn't do anything. In the case of `ListView` it's always a `StaticRenderer` which ignores the `collapsable` prop. I'm kind of leaning back towards adding the `collapsableChildren` prop to `View` that ensures that all it's children map to a native `View` and don't get optimized out.

Here's a simple example to show why adding  the `collapsable` props to `ScrollView` children in JS doesn't work:
``` js
const CustomView = ({ text }) => <View><Text>{text}</Text></View>;

<ScrollView>
   // Works, view gets rendered with `collapsable=false`
  <View />
  // Doesn't work because the collapsable prop gets passed to `CustomView` but is not 
  // passed down to it's `View` child
  <CustomView text="hello" />
</ScrollView>
```

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java (hunk 4 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java]: Got it working with https://github.com/facebook/react-native/pull/9456/files#diff-a73c2a0c22eceb34cf86aaede40f931fR192. Tested that it does indeed work by animating the transform prop in a clipped view and I can confirm that you were right and it did not work before but now works with this fix.

**Comment by nihgwu** [comment]
what's going on, waiting for this feature to be merged😁

**Comment by janicduplessis** [comment]
Just need to figure out a few implementation details with @astreet, shouldn't take too long :)

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java (hunk 3 above)
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Ok, I understand the problems you're having. Ideally we'd have some way to refer to the child views other than by index (since it isn't stable), but that doesn't exist right now and I don't know a good way to do it. I'm fine doing collapsableChildren false in this case then. (Actually, another option would be a create a custom StickyHeader native view class that ScrollView knows about and can do instanceof checks in native... but that would require a change in API)

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Let's go with collapsableChildren, this will allow to keep the same API and could even be useful in the future for other cases where we want an index based API like stickyHeaderIndices.

**Comment by facebook-github-bot-2** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/8c39c9b73f931c881311664fe6ea1d7543d306d1..72ce773e37530d90bfdc7cef5bb70f3f9d797f41)

**Comment by facebook-github-bot-5** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/72ce773e37530d90bfdc7cef5bb70f3f9d797f41..4f09d2067f3a4ea32062f00aa20f79f4ed680b43)

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java [hunk 5]
```diff
@@ -435,15 +439,39 @@ private void transitionLayoutOnlyViewToNativeView(
...
+    if (props.hasKey(ViewProps.COLLAPSABLE_CHILDREN) && !props.getBoolean(ViewProps.COLLAPSABLE_CHILDREN, true)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private static boolean isCollapsable(@Nullable ReactStylesDiffMap props) {
+    if (props == null) {
+      return false;
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: This should return true for consistency with collapsableChildren (and to maintain current behavior)

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java [hunk 6]
```diff
@@ -435,15 +439,39 @@ private void transitionLayoutOnlyViewToNativeView(
     mTagsWithLayoutVisited.clear();
   }
 
-  private static boolean isLayoutOnlyAndCollapsable(@Nullable ReactStylesDiffMap props) {
+  private static boolean isCollapsableChildren(@Nullable ReactStylesDiffMap props) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: I think we can improve the name here... it's a bit confusing to me. Maybe `areChildrenCollapsable`?

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java [hunk 7]
```diff
@@ -435,15 +439,39 @@ private void transitionLayoutOnlyViewToNativeView(
...
 
+    return true;
+  }
+
+  private static boolean isLayoutOnlyAndCollapsable(@Nullable ReactStylesDiffMap props) {
+    if (props == null) {
+      return true;
+    }
+
+    if (!isCollapsable(props) || !isCollapsableChildren(props)) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: For the second part of this `if`: the behavior here is that if a view has collapsableChildren=false set on it, it will also not be collapsable itself. Is that necessary?

**Comment by astreet** [comment]
Looks good, can you just clean up the collapsableChildren comments?

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java (hunk 7 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyOptimizer.java]: Haven't tested a case where that would happen but I'm pretty sure it would cause issues if we optimize out a view that we want to make sure it's child exists and map to the same index as Views in react js.

**Comment by facebook-github-bot-4** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/4f09d2067f3a4ea32062f00aa20f79f4ed680b43..f750faf7970fbb9549cc7772c2b98290cd4b8152)

**Comment by janicduplessis** [comment]
@astreet Yeah that naming was really bad, rename the prop to `collapseChildren` and methods to `shouldCollapseChildren`.

**Comment by astreet** [comment]
@facebook-github-bot shipit

Great, thanks!

**Comment by astreet** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-9** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3827366) to review internal test results.

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java [hunk 8]
```diff
@@ -7,13 +7,11 @@
  * of patent rights can be found in the PATENTS file in the same directory.
  */
 
-package com.facebook.react.views.view;
+package com.facebook.react.uimanager;
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java]: Was there a reason this was moved? There are internal classes in FB that are referencing this class and that's what's causing this patch to fail to land. It'd be easiest if we don't move this if we don't need to.

Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java]: I moved it to avoid a circular reference of dependencies when building with buck.

**Comment by facebook-github-bot-1** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java (hunk 8 above)
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java]: ah, I see. How about I do the refactor internally and you rebase this diff once it lands?

**Comment by astreet** [comment]
I put up the diff for the refactor and it's landing now. Keep an eye out for it in a few hours (hopefully it lands without problems) :)

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java (hunk 8 above)
Comment (janicduplessis) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/ReactClippingViewGroup.java]: Sure sounds good!

**Comment by brentvatne** [comment]
@astreet - it would seem that it did not land :( anything we can do to help?

**Comment by astreet** [comment]
Sorry, I was hitting some transient infra failures trying to land it. It's now landed as https://github.com/facebook/react-native/commit/e2cf37a5a60e4031114c58d04cc1379007584cf6 :)

**Comment by facebook-github-bot-2** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/f750faf7970fbb9549cc7772c2b98290cd4b8152..d0a8f12bf138a06031cc6f8a3fd4786121df5c6f) - [changes since last import](https://github.com/facebook/react-native/pull/9456/files/f750faf7970fbb9549cc7772c2b98290cd4b8152..d0a8f12bf138a06031cc6f8a3fd4786121df5c6f)

**Comment by janicduplessis** [comment]
@astreet Rebased on top of your diff

**Comment by astreet** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-3** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3827366) to review internal test results.

**Comment by facebook-github-bot-5** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by astreet** [comment]
Diff didn't land because of the ScrollView ctor change: we have internal subclasses of that class. I sent out a refactor that will hopefully land soon.

**Comment by janicduplessis** [comment]
👍 I saw that some ScrollView instrumentation tests are failling too on circleci, gonna take a look at it.

**Comment by astreet** [comment]
Cool, scrollview refactor landed as https://github.com/facebook/react-native/commit/2cf2fdbc04bc5309f6942ed464ad61cd06a26ce8

**Comment by facebook-github-bot-2** [comment]
@janicduplessis updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9456/files/d0a8f12bf138a06031cc6f8a3fd4786121df5c6f..51e0acceb5d451592088d3127752d1c864e38bf8) - [changes since last import](https://github.com/facebook/react-native/pull/9456/files/d0a8f12bf138a06031cc6f8a3fd4786121df5c6f..51e0acceb5d451592088d3127752d1c864e38bf8)

**Comment by janicduplessis** [comment]
@astreet Ok should be all good now, tests were failing because the change in NativeViewHierachyOptimizer caused it not to optimize out view that their parent is null. Fixed in 51e0acc. Now tests are passing locally.

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-5** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3827366) to review internal test results.

**Comment by nihgwu** [comment]
OMG, failing still?

**Comment by janicduplessis** [comment]
Landed now :D

**Comment by kenma9123** [comment]
Been waiting for this. Thank you!

**Comment by ide** [comment]
Reverted in https://github.com/facebook/react-native/commit/d0d1712851d05595b13cf8cc558cec7894ec24f5

@astreet Could you paste stack traces / the reason for the revert here so we can fix it?

**Comment by astreet** [comment]
Yep, sorry for posting after the fact, this got reverted after I left for the day. We hit a couple problems after this landed:

1. The next-gen version of the RN Android rendering stack (nodes) doesn't use `ReactViewGroup` (and therefore doesn't use `ReactViewManager`). This was fixed in https://github.com/facebook/react-native/commit/5c3f9547c6b1812c5d73def4000d748bd2fefb21 (but this was also reverted as part of the revert of the larger stickyheader commit). When we re-land, we should incorporate this fix.
2. We got this bug report from adsmanager that bisects to this diff ("our scrollview has 2 views, 1 on top of another, lets say A on top of B. We can swipe A horizontally to reveal B. With this diff, B is always shown instead of A"). I haven't had a chance to try to repro, but I imagine this has to do with the drawing order changes. I'll try to provide more information when I get a chance to look into it, but I had them go ahead and revert since they have a release coming up in a few days.

**Comment by astreet** [comment]
Looking into this more, one of the things that broke was that ListView automatically enables sticky headers for section headers. Some of the code in adsmanager for android apparently doesn't properly identify what things are section headers vs. normal rows and we now have random rows sticking to the top ha.

I think before this diff lands, we need to have whether stick headers are enabled for ListView customizable. And, given the design patterns on the two platforms, it should probably be enabled by default on iOS (as it is currently) and DISABLED by default on Android. Blindly copying design patterns from iOS to Android, like with sticky headers, is one of those things that identifies your framework as a lowest-common denominator crossplatform solution which we don't want. That being said, I think there are legit use cases for it, I just don't want it to become a 'thing' for RN apps on Android.

**Comment by astreet** [comment]
I haven't been able to get a repro for the rendering issue that was described to me (where B renders over A), but I have a feeling it's because stickyHeaderIndices is causing us to identify the wrong View as being sticky somehow.

I'd really feel a lot better about this API if, instead of passing in indices, we wrapped the views that are supposed to be sticky in a custom native view that the ScrollView can identify. That way, we aren't hurt by collapsable optimizations and don't need to make sure stickyHeaderIndices and the ListView dataSource stay in sync.

**Comment by nihgwu** [comment]
__ScrollView Sticky Header on Android__ is really important to me, I'm building a cross-platform app, it's really ugly without sticky header on Android, I've made a js workaround but far from satisfaction, so I build RN from source to use this feature, and it works great, although there are still some bugs but it's fine in my case.

I think it's deserve to be kept in the master branch for further improvement, while providing an option to disable it on Android for consistency, e.g. `stickyHeaderIndices={[]}`

**Comment by nihgwu** [comment]
And I've noticed the random sticking issue in `UIExplorer`, the example list in the drawer and `React Native Examples` has section headers but stick unexpectedly, while it's works well in `ListViewPagingExample`

**Comment by janicduplessis** [comment]
@astreet I think I know why tbe rendering but you describe happens. Gonna have a look at this as well as the issues @nihgwu found out today.

I also agree that we should make this disabled by default on android and add a flag to listview to enable it. Thisatches better what is common on the platform and will also avoid making this a breaking change.

**Comment by janicduplessis** [comment]
Also for using a custom native view for headers I agree this would be a better and more reliable API but it should still be possible to make indices work, it's just a matter of fixing a few remaining bugs then it will be stable.

**Comment by nihgwu** [comment]
I add a `renderStickyHeader` in my workaround `ListView` for Android implemented in js , but it looks silly as there is already `renderSectionHeader` means to render sticky header on iOS

**Comment by janicduplessis** [comment]
I'll close this PR and open a new one since we already shipped this, not sure if the bot will work properly.

**Comment by antoinerousseau** [comment]
ref: https://github.com/facebook/react-native/issues/2700
